x/common/error.go

```diff
@@ -3,45 +3,180 @@ package common
 import (
 	"errors"
 	"fmt"
+	"runtime/debug"
 )
 
-// ToError converts a value to error if if (1) is a string, (2) has a String() function
-// or (3) is already an error.
-func ToError(v any) error {
+// TryCatch is an implementation of the try-catch block from languages like C++ and JS.
+// Given a 'callback' function, TryCatch defers and recovers from any panics or
+// errors, allowing one to safely handle multiple panics in succession.
+//
+// Typically, you'll write something like: `err := TryCatch(aRiskyFunction)()`
+//
+// Usage example:
+//
+//	var calmPanic error = TryCatch(func() {
+//	  panic("something crazy")
+//	})()
+//	fmt.Println(calmPanic.Error()) // prints "something crazy"
+//
+// Note that TryCatch doesn't return an error. It returns a function that returns
+// an error. Only when you call the output of TryCatch will it "feel" like a
+// try-catch block from other languages.
+//
+// This means that TryCatch can be used to restart go routines that panic as well.
+func TryCatch(callback func()) func() error {
+	return func() (err error) {
+		defer func() {
+			if panicInfo := recover(); panicInfo != nil {
+				err = fmt.Errorf("%v, %s", panicInfo, string(debug.Stack()))
+				return
+			}
+		}()
+		callback() // calling the decorated function
+		return err
+	}
+}
+
+// ToError converts a value to an error type if it:
+// (1) is a string,
+// (2) has a String() function
+// (3) is already an error.
+// (4) or is a slice of these cases
+// I.e., the types supported are:
+// string, []string, error, []error, fmt.Stringer, []fmt.Stringer
+//
+// The type is inferred from try catch blocks at runtime.
+func ToError(v any) (out error, ok bool) {
 	if v == nil {
-		return nil
+		return nil, true
 	}
 	switch v := v.(type) {
 	case string:
-		return errors.New(v)
+		return errors.New(v), true
 	case error:
-		return v
+		return v, true
 	case fmt.Stringer:
-		return errors.New(v.String())
+		return errors.New(v.String()), true
+	case []string:
+		return toErrorFromSlice(v)
+	case []error:
+		return toErrorFromSlice(v)
+	case []fmt.Stringer:
+		return toErrorFromSlice(v)
 	default:
-		panic(fmt.Errorf("invalid type: %T", v))
+		// Cases for duck typing at runtime
+
+		// case: error
+		if tcErr := TryCatch(func() {
```

Comment:
That's how I thought it would work at first too. The reason it skips those blocks is that `any` and `interface{}` are actual types. They don't actual mean "any type" but rather "an object of type `any`". 

Let's say we pass in an argument, `arg`, that fits the `fmt.Stringer` interface like `sdk.Coin` or `sdk.AccAddress`. When `arg` enters this function, it won't actually have the type of `fmt.Stringer`, so it will pass all of the case statements and end up in default. 

Then, when it reaches the `arg := arg.(fmt.Stringer)` block, it will enter the function with `fmt.Stringer` as its value for `arg.(type)`. Since `arg` actually *is* an instance of that fits this interface, the `TryCatch` will run smoothly. 

However, it has to guess by manually attempting to cast as each type by panicking if it doesn't make sense, then defer, then recover, and finally move to the next `TryCatch`. 

This basically shows why Python is so slow.